### PR TITLE
Add 'axis' option for 'tf.boolean_mask()'

### LINKED ISSUE #9721: Feature request: Add 'axis' option for 'tf.boolean_mask()'
`tf.boolean_mask()` is extremely useful when it comes to training.

However, the present `tf.boolean_mask()` does not have an `axis` option, thus masking some specific dimension of an array could be troublesome. 

Appreciate your consideration.

### PR DESCRIPTION
**Pull request by yongtang**
This fix tries to address #9721 where it was not possible to pass an
'axis' option for 'tf.boolean_mask()'.

This fix fixes #9721.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

File: tensorflow/python/ops/array_ops.py [hunk 1]
```diff
@@ -1084,6 +1084,7 @@ def boolean_mask(tensor, mask, name="boolean_mask"):
   Args:
     tensor:  N-D tensor.
     mask:  K-D boolean tensor, K <= N and K must be known statically.
+    axis: The axis in `tensor` to mask from. Defaults to the first dimension.
```
Comment (josh11b) [inline tensorflow/python/ops/array_ops.py]: What if mask isn't a 1-D tensor?

File: tensorflow/python/ops/array_ops.py [hunk 2]
```diff
@@ -1066,7 +1066,7 @@ def concat(values, axis, name="concat"):
                                   name=name)
 
 
-def boolean_mask(tensor, mask, name="boolean_mask"):
+def boolean_mask(tensor, mask, axis=None, name="boolean_mask"):
```
Comment (josh11b) [inline tensorflow/python/ops/array_ops.py]: For backwards compatibility, all new arguments must be last (in addition to having defaults).

**Review by josh11b — COMMENTED** [review]
(API review.)

File: tensorflow/python/ops/array_ops.py (hunk 2 above)
Comment (yongtang) [inline tensorflow/python/ops/array_ops.py]: Thanks @josh11b. The PR has been updated with the suggested changes.

File: tensorflow/python/ops/array_ops.py (hunk 1 above)
Comment (yongtang) [inline tensorflow/python/ops/array_ops.py]: @josh11b If mask isn't 1-D then it should work as well. I added an additional test case for this.

**Comment by yongtang** [comment]
@josh11b Thanks for the review. The PR has been updated. Please take a look.

File: tensorflow/python/ops/array_ops.py [hunk 3]
```diff
@@ -1085,6 +1085,7 @@ def boolean_mask(tensor, mask, name="boolean_mask"):
     tensor:  N-D tensor.
     mask:  K-D boolean tensor, K <= N and K must be known statically.
     name:  A name for this operation (optional).
+    axis: The axis in `tensor` to mask from. Defaults to the first dimension.
```
Comment (josh11b) [inline tensorflow/python/ops/array_ops.py]: You should clarify the documentation. In particular you should mention that it must be an integer, and please update lines 1080-1082. Ideally in addition `axis` would also be allowed to be a list of length K to be consistent with other `axis` arguments in the API, e.g. https://www.tensorflow.org/api_docs/python/tf/reduce_sum .

**Review by josh11b — DISMISSED** [review]
(for API review)

File: tensorflow/python/ops/array_ops.py (hunk 3 above)
Comment (langmore) [inline tensorflow/python/ops/array_ops.py]: `boolean_mask` already can work as an n-dimensional mask.  Currently if `mask.shape.ndims = K`, it masks over the first `K` dimensions of `tensor`.  So the docstring should be changed to reflect the new behavior if e.g. `axis != None` and `mask` is multi-dimensional.

Comment (langmore) [inline tensorflow/python/ops/array_ops.py]: `boolean_mask(tensor, mask)` is supposed to act like numpy's `my_array[my_mask]`, which it currently does.  What you're proposing here looks like [`numpy.compress`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.compress.html#numpy.compress)

So it seems like this functionality should be part of an `array_ops.compress`.

Comment (josh11b) [inline tensorflow/python/ops/array_ops.py]: So numpy.compress only allows 1-D boolean masks, which makes the choice of only using an integer for the axis straightforward.  This op allows K-D boolean masks, which raises the question of whether the axis should be (allowed to be) a length-K vector.

**Comment by jhseu** [comment]
@yongtang Any updates? Thanks!

**Comment by yongtang** [comment]
@josh11b @langmore Sorry for missing the GitHub notification and late reply. I have rebased and updated PR with the docstrings addressed based on the comments.

Please take a look and let me know if there are any issues.

**Comment by drpngx** [comment]
@langmore @josh11b any additional thoughts?

Jenkins, test this please.

**Comment by sb2nov** [comment]
@langmore @josh11b any additional thoughts?

**Comment by josh11b** [comment]
This is good for api review.

**Comment by yongtang** [comment]
Thanks @josh11b for the review. The API goldens have been updated so that `//tensorflow/tools/api/tests:api_compatibility_test` could be passed.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Review by martinwicke — APPROVED** [review]

**Comment by martinwicke** [comment]
Ignoring Kokoro -- looks like tool failure to me.

**Pull request merged**